log: allow Finishing an EventLog concurrently with events

RaduBerinde opened the pull request:
(Please ignore the first commit, it's a separate PR)

The logging code now "owns" the EventLog and synchronizes access to it. We now
allow logging calls concurrent with finishing the eventlog.

Verified using

```
make stressrace PKG=./gossip TESTS=TestClientGossip TESTFLAGS='--verbosity 2'
```

Fixes #8910.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8913)

<!-- Reviewable:end -->

tamird commented:
Which PR does this depend on? I'd rather review this after dependent changes have gone in.

RaduBerinde commented:
#8850

tamird commented:
Reviewed 4 of 4 files at r2.
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/queue.go, line 367 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-KQNaFn8GO2sIpSFRtt_:-KQNaFn8GO2sIpSFRtta:b-o0xgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d870880ac5fe4339f14fe05aad88f2324bbc495b/storage/queue.go#L367)):*
> ```Go
> 	defer bq.mu.Unlock()
> 
> 	if bq.mu.stopped {
> ```

is the `stopped` field still needed?

---

*[util/log/trace.go, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-KQNaoz8SNFVvLy-sstF:-KQNaoz8SNFVvLy-sstG:b-ixdod5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d870880ac5fe4339f14fe05aad88f2324bbc495b/util/log/trace.go#L82)):*
> ```Go
> // the current ctxEventLog. If neither (or the Span is NoopTracer), returns
> // false.
> func getSpanOrEventLog(ctx context.Context) (opentracing.Span, *ctxEventLog,
> ```

can you wrap in accordance with STYLE.md?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/queue.go, line 367 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-KQNaFn8GO2sIpSFRtt_:-KQNbjLENufeq5qcXOAs:b89hgul) ([raw file](https://github.com/cockroachdb/cockroach/blob/d870880ac5fe4339f14fe05aad88f2324bbc495b/storage/queue.go#L367)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is the `stopped` field still needed?
</details>
I think it still makes sense to prevent the use of the queues after they were stopped (#8561), even if there is no longer a race around the eventlog.

I added a cleanup commit where I am moving `ctx` out of `mu`.

---

*[util/log/trace.go, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-KQNaoz8SNFVvLy-sstF:-KQNc2mzjAMEsV4_s5kO:b-izv4ai) ([raw file](https://github.com/cockroachdb/cockroach/blob/d870880ac5fe4339f14fe05aad88f2324bbc495b/util/log/trace.go#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you wrap in accordance with STYLE.md?
</details>
Oops, the newline was accidental.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/queue.go, line 367 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-KQNaFn8GO2sIpSFRtt_:-KQOLe4tN6YHUdbDpWYf:b-5hwf0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d870880ac5fe4339f14fe05aad88f2324bbc495b/storage/queue.go#L367)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I think it still makes sense to prevent the use of the queues after they were stopped (#8561), even if there is no longer a race around the eventlog.
> 
> I added a cleanup commit where I am moving `ctx` out of `mu`.
</details>
Interestingly that commit doesn't seem to change any of the locking.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-:-KQOLtxzziix8UMOAF76:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. Commit message should refer to #8905, not #8910, right?

---

Reviewed 1 of 4 files at r2, 2 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[util/log/trace.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-KQOkdWeD_16b8RHBXMA:-KQOkdWeD_16b8RHBXMB:b54s37d) ([raw file](https://github.com/cockroachdb/cockroach/blob/733acb969dc65299eaed91e0fa30d564108a9719/util/log/trace.go#L42)):*
> ```Go
> // and event calls to go to the EventLog. The current context must not have an
> // existing open span.
> func withEventLogInt(ctx context.Context, eventLog trace.EventLog) context.Context {
> ```

`Int` == `Internal`? That's not a suffix we use much. I'd spell the word out in full. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-:-KQOk2dQTlG4W9quNNGe:b-rfgwip)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
log: allow Finishing an EventLog concurrently with events

The logging code now "owns" the EventLog and synchronizes access to it. We now
allow logging calls concurrent with finishing the eventlog.

Verified using
```
make stressrace PKG=./gossip TESTS=TestClientGossip TESTFLAGS='--verbosity 2'
```

Fixes #8905.

RaduBerinde committed:
storage: clean up queue.ctx usage

The only reason `ctx` was protected by the mutex was because of the embedded
EventLog. Now that we internally synchronize access to it, we no longer need
this complication; moving `ctx` out of `mu`.

RaduBerinde commented:
Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[util/log/trace.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913#-KQOkdWeD_16b8RHBXMA:-KQQzjGIa3ApnPxN54rd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/733acb969dc65299eaed91e0fa30d564108a9719/util/log/trace.go#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `Int` == `Internal`? That's not a suffix we use much. I'd spell the word out in full. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8913)*
<!-- Sent from Reviewable.io -->
